shade-memcached/src/main/scala/shade/memcached/Codec.scala

```diff
@@ -0,0 +1,201 @@
+
+  /** Decodes byte arrays with flags, as retrieved by the Memcached client,
+    * into the value it represents.
+    */
+  def decode(data: CachedData): T
+
+  /** Get the maximum size of objects handled by this codec. */
+  def getMaxSize: Int
+}
+
+object Codec extends DefaultCodecs
+
+abstract class DefaultCodecs extends DefaultCodecsLevel0 {
+  import java.lang.{Float => JvmFloat, Double => JvmDouble}
+
+  /** Default codec for `Int`. */
+  implicit object IntBinaryCodec extends GenericIntCodec[Int](
+    flags = 2 << 8, // SerializingTranscoder.SPECIAL_INT
+    toInt = (v: Int) => v,
+    fromInt = (v: Int) => v
+  )
+
+  /** Default codec for `Long`. */
+  implicit object LongBinaryCodec extends GenericLongCodec[Long](
+    flags = 3 << 8, // SerializingTranscoder.SPECIAL_LONG
+    toLong = (v: Long) => v,
+    fromLong = (v: Long) => v
+  )
+
+  /** Default codec for `Float`. */
+  implicit object FloatBinaryCodec extends GenericIntCodec[Float](
+    flags = 6 << 8, // SerializingTranscoder.SPECIAL_FLOAT
+    toInt = JvmFloat.floatToRawIntBits,
+    fromInt = JvmFloat.intBitsToFloat
+  )
+
+  /** Default codec for `Double`. */
+  implicit object DoubleBinaryCodec extends GenericLongCodec[Double](
+    flags = 7 << 8, // SerializingTranscoder.SPECIAL_DOUBLE
+    toLong = JvmDouble.doubleToRawLongBits,
+    fromLong = JvmDouble.longBitsToDouble
+  )
+
+  /** Default codec for `Byte`. */
+  implicit object ByteBinaryCodec extends Codec[Byte] {
+    final val FLAGS = 5 << 8 // SerializingTranscoder.SPECIAL_BYTE
+
+    def asyncDecode(d: CachedData): Boolean = false
+
+    def encode(value: Byte): CachedData = {
+      val bytes = packedUtils.encodeByte(value)
+      new CachedData(FLAGS, bytes, getMaxSize)
+    }
+
+    def decode(data: CachedData): Byte =
+      data.getData match {
+        case null => 0
+        case bytes =>
+          packedUtils.decodeByte(bytes)
+      }
+
+    def getMaxSize: Int =
+      MAX_SIZE
+  }
+  
+  /** Default codec for `Boolean`. */
+  implicit object BooleanCodec extends Codec[Boolean] {
+    // SerializingTranscoder.SPECIAL_BOOLEAN
+    final val FLAGS = 1 << 8
+
+    def asyncDecode(d: CachedData): Boolean = false
+
+    def encode(value: Boolean): CachedData = {
+      val bytes = packedUtils.encodeBoolean(value)
+      new CachedData(FLAGS, bytes, getMaxSize)
+    }
+
+    def decode(data: CachedData): Boolean =
+      data.getData match {
+        case null => false
+        case bytes =>
+          packedUtils.decodeBoolean(bytes)
+      }
+
+    def getMaxSize: Int =
+      MAX_SIZE
+  }
+}
+
+private[memcached] abstract class DefaultCodecsLevel0 {
+  /** Returns a [[Codec]] reference that can serialize and 
+    * deserialize anything. 
+    */
+  implicit def serializingCodec[A]: Codec[A] =
+    SerializingCodecAnyRef.asInstanceOf[Codec[A]]
+  
+  /** Reusable reference for [[serializingCodec]]. */
+  private object SerializingCodecAnyRef extends Codec[Any] {
```

Comment:
I wonder if it might be worthwhile to add in Shapeless to build an auto generic codec that can be tried before we fall back to something that relies on Java serialisation.